### PR TITLE
We need to sort mounts so that one mount does not over mount another.

### PR DESCRIPTION
**Pull request by rhatdan**
Currenly we add mounts from images, volumes and internal.
We can accidently over mount an existing mount.  This patch sorts the mounts
to make sure a parent directory is always mounted before its content.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
It fixes the mount order but unmounting is now an issue: `device or resource busy`

File: libpod/container_internal_linux.go [hunk 1]
```diff
@@ -248,9 +250,38 @@ func (c *Container) generateSpec(ctx context.Context) (*spec.Spec, error) {
...
+	// Mounts need to be sorted so paths will not cover other paths
+	mounts := sortMounts(g.Mounts())
+	g.ClearMounts()
+	for _, m := range mounts {
+		g.AddMount(m)
+	}
 	return g.Config, nil
 }
 
+type byDestination []spec.Mount
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal_linux.go]: Could this sort code be moved to util.go or some such?

**Comment by rhatdan** [comment]
@vrothberg Could you give me a command line where this causes an issue?

**Comment by vrothberg** [comment]
> @vrothberg Could you give me a command line where this causes an issue?

@rhatdan, see below:
`podman run --rm -v `pwd`/install.md:/config/install.md uchiwa/uchiwa:latest  ls /config`

**Comment by rhatdan** [comment]
@vrothberg, I am in a car with limited bandwith, could you check if the current patch set works for your use case.

**Comment by vrothberg** [comment]
> @vrothberg, I am in a car with limited bandwith, could you check if the current patch set works for your use case.

Don't hack and drive, sure thing. Unfortunately, I still see the error. The latest changes make a lot of sense though and I expected them to work. I'll dig a bit deeper.

**Comment by vrothberg** [comment]
@rhatdan Those mounts are stored in the run-time spec and it looks like we're not keeping that in memory. I am not very familiar with those parts of the code but I think should keep the rspec as a field for the Container.

`Container.Config.Mounts` will only point to `[...]/userdata` at that point.

**Comment by mheon** [comment]
Do we even need `container.config.Mounts`? It's only ever used for holding the path to /dev/shm

**Comment by rhatdan** [comment]
Yes I think so, it keeps track of tmpfs mounted to be shared between containers.

**Comment by mheon** [comment]
It really doesn't belong in config, given it's not statically managed. And we're not doing any tracking to see if other containers are using our SHM before unmounting (we might want to look into doing that)

We should look into expanding the bind-mount handling we have in State now to include tmpfs creation (and mount options) and combine it with that. That should be a separate PR, though.

**Comment by rhatdan** [comment]
@vrothberg Sadly took me a day to figure it out.  The issue you were seeing was caused by the Image Volumes being mounted as "shared" propagation.  Then when podman added the volume mount into the image volume the mount point showed up on the host mount namespace as well as in the container.  Changing volume images mounts to "private" fixes the problem.

**Comment by mheon** [comment]
LGTM

File: libpod/container_internal_linux.go [hunk 2]
```diff
@@ -248,6 +248,13 @@ func (c *Container) generateSpec(ctx context.Context) (*spec.Spec, error) {
 		g.SetLinuxCgroupsPath(cgroupPath)
 	}
 
+	// Mounts need to be sorted so paths will not cover other paths
+	mounts := sortMounts(g.Mounts())
+	g.ClearMounts()
+	for _, m := range mounts {
+		fmt.Println("DAN Mount", m)
```
Comment (vrothberg) [inline libpod/container_internal_linux.go]: I guess that's a debugging artifact (although it would be a funny easter egg).

**Comment by mheon** [comment]
@rhatdan Fix the debugging artifact and I'll merge

**Comment by mheon** [comment]
Also have lint warnings

**Comment by rhatdan** [comment]
@rh-atomic-bot r=mheon

**Pull request closed without merging**